presto-hive/src/test/java/com/facebook/presto/hive/TestEventListenerWithExchangeMaterialization.java

```diff
@@ -0,0 +1,264 @@
+
+        assertFalse(runtimeOptimizedStages.isPresent());
+        assertTrue(queryCompletedEvent.getMetadata().getRuntimeOptimizedStages().isEmpty());
+
+        // Now, the following query should have two optimized joins in a single stage (both on the same nationkey), therefore expect only one optimized stage: 1.
+        runQueryAndWaitForEvents("SELECT supplier.phone, regionkey, custkey FROM nation INNER JOIN supplier ON supplier.nationkey=nation.nationkey INNER JOIN customer ON nation.nationkey=customer.nationkey", expectedEvents);
+        queryCreatedEvent = getOnlyElement(generatedEvents.getQueryCreatedEvents());
+        queryCompletedEvent = getOnlyElement(generatedEvents.getQueryCompletedEvents());
+        queryMetadata = queryCompletedEvent.getMetadata();
+        runtimeOptimizedStages = queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(new QueryId(queryCreatedEvent.getMetadata().getQueryId())).getRuntimeOptimizedStages();
+
+        assertTrue(runtimeOptimizedStages.isPresent() && !queryMetadata.getRuntimeOptimizedStages().isEmpty());
+        assertTrue(queryMetadata.getRuntimeOptimizedStages().size() == 1 && runtimeOptimizedStages.get().size() == 1);
+        assertEquals(queryMetadata.getRuntimeOptimizedStages(), runtimeOptimizedStages.get().stream()
+                .map(stageId -> String.valueOf(stageId.getId()))
+                .collect(toImmutableList()));
+        assertEquals(queryMetadata.getRuntimeOptimizedStages().get(0), "1");
+
+        // Now, the following query should have two runtime optimized stages: 1 and 4, corresponding to the two join operations (on regionkey and nationkey respectively).
+        runQueryAndWaitForEvents("WITH natreg AS (SELECT nation.regionkey, nationkey, region.name FROM region INNER JOIN nation ON nation.regionkey=region.regionkey) SELECT phone, regionkey FROM natreg INNER JOIN supplier ON supplier.nationkey=natreg.nationkey", expectedEvents);
+        queryCreatedEvent = getOnlyElement(generatedEvents.getQueryCreatedEvents());
+        queryCompletedEvent = getOnlyElement(generatedEvents.getQueryCompletedEvents());
+        queryMetadata = queryCompletedEvent.getMetadata();
+        runtimeOptimizedStages = queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(new QueryId(queryCreatedEvent.getMetadata().getQueryId())).getRuntimeOptimizedStages();
+
+        assertTrue(runtimeOptimizedStages.isPresent() && !queryMetadata.getRuntimeOptimizedStages().isEmpty());
+        assertTrue(queryMetadata.getRuntimeOptimizedStages().size() == 2 && runtimeOptimizedStages.get().size() == 2);
+        assertEquals(queryMetadata.getRuntimeOptimizedStages(), runtimeOptimizedStages.get().stream()
+                .map(stageId -> String.valueOf(stageId.getId()))
+                .collect(toImmutableList()));
+        assertEquals(ImmutableSet.copyOf(queryMetadata.getRuntimeOptimizedStages()), ImmutableSet.of("1", "4"));
+    }
+
+    static class TestingEventListenerPlugin
```

Comment:
Is this a copy of what's in TestEventListenerPlugin?  You should be able to use those classes directly, e.g. like TestCompletedEventWarnings (maybe worth extracting all these TestingEventListener things to a top level class since it's used across multiple test classes

Reply:
Discussed offline that as this test has to be under presto-hive to naturally enable exchange materialization, it is ok to leave it as it is because it is not directly accessible to TestEventListenerPlugin class.